libs/ssl-config/src/test/java/org/elasticsearch/common/ssl/PemKeyConfigTests.java

```diff
@@ -70,6 +70,7 @@ public void testBuildKeyConfigFromPkcs8PemFilesWithoutPassword() throws Exceptio
     }
 
     public void testBuildKeyConfigFromPkcs8PemFilesWithPassword() throws Exception {
+        assumeFalse("Can't run in a FIPS JVM, PBE KeySpec is not available", inFipsJvm());
```

Comment:
This is probably a dumb question. This is not possible in FIPS mode because the key is genereated using the PBEKeySpec?

Reply:
This is not a dumb question. I realize that many times in the FIPS journey the comments and documentation I added made sense to me ( and sometimes not for long after the time when I wrote them ) but are not proper explanations. 

To start with, we are limited to using only the algorithms defined in PKCS#5 1.5 because of a JDK bug ( that will hopefully be solved in JDK15 ) see :https://github.com/elastic/elasticsearch/issues/32021.

When not in FIPS mode, `javax.crypto.SecretKeyFactory#generateSecret` uses `com.sun.crypto.provider.PBEKeyFactory#engineGenerateSecret` which can happily consume a PBEKeySpec that has no salt because it uses `com.sun.crypto.provider.PBEKey` which doesn't have a salt ( the salt _is_ used, but later on by the `Cipher` ) . 

When in FIPS mode, `javax.crypto.SecretKeyFactory#generateSecret` uses `org.bouncycastle.jca.jce.provider.ProvPBE$PBKDF1#engineGenerateSecret` that asserts that the PBEKeySpec has a salt and fails if not. 

Now we could restructure our `PemUtils` code to get the `PbeParameterSpec` with `encryptedPrivateKeyInfo.getAlgParameters().getParameterSpec(AlgorithmParameterSpec.class);` and then get the salt and iterations from there so that we could initialize a `PBEKeySpec` with the password, salt and iteration count so that it works in both SunJCE and BouncyCastle FIPS Provides _BUT_ here's the catch: 

BouncyCastle FIPS provider doesn't support the algorithms that are described in `PKCS#5 v1.5` encryption . ( `PBE-MD2-DES` and `PBE-MD5-DES` ), so it can't parse them to an `AlgorithmParameterSpec`. 


In summary `PBE KeySpec is not available` is an understatement that made sense 2 years ago but in truth should have been described in detail in an issue that we can keep track of this. I will do this now, thanks for asking!

Reply:
Shouldn't our FIPS instructions say that PKCS#8 password protected keys are not supported then?
Ironically, this bug was originally reported by someone trying to set up a FIPS 140-2 compliant cluster.

Reply:
To be fair, we don't support PKCS#5 v2 encrypted PKCS# encoded PEM keys at all and  we don't support PKCS#5 v1.5 encrypted PKCS# encoded PEM keys in FIPS mode. At that time when we discovered this, the issue seemed to cover this since it affected the encryption that is prefered ( v2 ) irrespective to FIPS but you're right this should be marked as a limitation.